collaborative-plonk/src/round3.rs

```diff
@@ -0,0 +1,602 @@
+    pub(super) alpha2: P::ScalarField,
+}
+
+pub(super) struct FinalPolys<T, P: Pairing>
+where
+    for<'a> T: PrimeFieldMpcProtocol<P::ScalarField>,
+{
+    pub(super) a: PolyEval<T, P>,
+    pub(super) b: PolyEval<T, P>,
+    pub(super) c: PolyEval<T, P>,
+    pub(super) z: PolyEval<T, P>,
+    pub(super) t1: FieldShareVec<T, P>,
+    pub(super) t2: FieldShareVec<T, P>,
+    pub(super) t3: FieldShareVec<T, P>,
+}
+impl<T, P: Pairing> FinalPolys<T, P>
+where
+    for<'a> T: PrimeFieldMpcProtocol<P::ScalarField>,
+{
+    fn new(
+        polys: Round2Polys<T, P>,
+        t1: FieldShareVec<T, P>,
+        t2: FieldShareVec<T, P>,
+        t3: FieldShareVec<T, P>,
+    ) -> Self {
+        Self {
+            a: polys.poly_eval_a,
+            b: polys.poly_eval_b,
+            c: polys.poly_eval_c,
+            z: polys.z,
+            t1,
+            t2,
+            t3,
+        }
+    }
+}
+
+impl<T, P: Pairing> Round3Challenges<T, P>
+where
+    for<'a> T: PrimeFieldMpcProtocol<P::ScalarField>,
+{
+    fn new(
+        round2_challenges: Round2Challenges<T, P>,
+        alpha: P::ScalarField,
+        alpha2: P::ScalarField,
+    ) -> Self {
+        Self {
+            b: round2_challenges.b,
+            beta: round2_challenges.beta,
+            gamma: round2_challenges.gamma,
+            alpha,
+            alpha2,
+        }
+    }
+}
+
+impl<T, P: Pairing> Round3<T, P>
+where
+    for<'a> T: PrimeFieldMpcProtocol<P::ScalarField>
+        + PairingEcMpcProtocol<P>
+        + FFTProvider<P::ScalarField>
+        + MSMProvider<P::G1>
+        + MSMProvider<P::G2>,
+    P::ScalarField: FFTPostProcessing,
+{
+    fn get_z1(domains: &Domains<P>) -> [P::ScalarField; 4] {
+        //TODO MOVE THIS THIS MUST BE A CONSTANT
+        let zero = P::ScalarField::zero();
+        let neg_1 = zero - P::ScalarField::one();
+        let neg_2 = neg_1 - P::ScalarField::one();
+        let root_of_unity = domains.roots_of_unity[2];
+        [zero, neg_1 + root_of_unity, neg_2, neg_1 - root_of_unity]
+    }
+
+    fn get_z2(domains: &Domains<P>) -> [P::ScalarField; 4] {
+        let zero = P::ScalarField::zero();
+        let two = P::ScalarField::one() + P::ScalarField::one();
+        let four = two.square();
+        let neg_2 = zero - two;
+        let root_of_unity = domains.roots_of_unity[2];
+        let neg2_root_unity = neg_2 * root_of_unity;
+        [
+            zero,
+            neg2_root_unity,
+            four,
+            P::ScalarField::zero() - neg2_root_unity,
+        ]
+    }
+
+    fn get_z3(domains: &Domains<P>) -> [P::ScalarField; 4] {
+        let zero = P::ScalarField::zero();
+        let two = P::ScalarField::one() + P::ScalarField::one();
+        let neg_eight = -(two.square() * two);
+        let root_of_unity = domains.roots_of_unity[2];
+        let two_root_unity = two * root_of_unity;
+        [zero, two + two_root_unity, neg_eight, two - two_root_unity]
+    }
+    fn compute_t(
```

Comment:
I think we should in general for this crate also document the methods that are non-pub, since this helps a lot when reading, even if only linking to the relevant section in the paper/circom source

Reply:
done